core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/DefaultPreprocessedValues.java

```diff
@@ -22,29 +23,29 @@ public DefaultPreprocessedValues(ProtocolBuilderNumeric builder) {
     return builder.seq(b -> {
       DRes<SInt> r = b.numeric().randomElement();
       DRes<SInt> inverse = b.advancedNumeric().invert(r);
-      ArrayList<DRes<SInt>> list = new ArrayList<>();
+      ArrayList<DRes<SInt>> list = new ArrayList<>(pipeLength + 2);
       list.add(inverse);
       list.add(r);
-      return () -> new IterationState(2, list);
-    }).whileLoop(
-        (state) -> state.round <= pipeLength + 1,
-        (seq, state) -> {
-          DRes<SInt> r = state.value.get(1);
-          DRes<SInt> last = state.value.get(state.value.size() - 1);
-          List<DRes<SInt>> values = new ArrayList<>(state.value);
-          values.add(seq.numeric().mult(last, r));
-          return () -> new IterationState(state.round + 1, values);
-        }).seq((seq, state) -> () -> state.value);
+      return () -> new IterationState(list);
+    }).whileLoop((state) -> state.values.size() <= pipeLength + 1, (seq, state) -> {
+      return seq.par(par -> {
+        DRes<SInt> last = state.values.get(state.values.size() - 1);
+        int limit = pipeLength + 2 - state.values.size();
+        List<DRes<SInt>> newValues = state.values.stream().skip(1).limit(limit)
+            .map(v -> par.numeric().mult(last, v)).collect(Collectors.toList());
+        state.values.addAll(newValues);
+        return () -> new IterationState(state.values);
+      });
+
+    }).seq((seq, state) -> () -> state.values);
   }
 
   private static final class IterationState {
```

Comment:
This is not needed anymore, remove it

Reply:
Which part?

Reply:
...

Trickster.

The entire class IterationState and can be replaced with the list

Reply:
Oh, you mean the iterationstate class? Sure I will get rid of that.